---
title: "06.2-cod-RNAseq-alignment-genome"
author: "Kathleen Durkin"
date: "2024-04-16"
always_allow_html: true
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true 
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

Code for aligning RNAseq data to reference genome, to be used on [Pacific cod RNAseq data](https://shedurkin.github.io/Roberts-LabNotebook/posts/projects/pacific_cod/2023_12_13_pacific_cod.html). 

- Raw reads found [here](https://owl.fish.washington.edu/nightingales/G_macrocephalus/30-943133806/)
- Trimmed reads: 
- Genome downloaded from [NCBI](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_031168955.1/), stored [here](https://owl.fish.washington.edu/halfshell/genomic-databank/GCF_031168955.1_ASM3116895v1_rna.fna) as a part of lab [genomic resources](https://robertslab.github.io/resources/Genomic-Resources/#gadus-macrocephalus-pacific-cod)

Note: Hisat2 pseudoalignment doesn't necessarily require input reads to be trimmed, provided they are of sufficient quality.

# Create a Bash variables file

This allows usage of Bash variables (e.g. paths to common directories) across R Markdown chunks.
```{r save-bash-variables-to-rvars-file, engine='bash', eval=TRUE}
{
echo "#### Assign Variables ####"
echo ""

echo "# Data directories"
echo 'export cod_dir=/home/shared/8TB_HDD_02/shedurkin/project-cod-temperature'
echo 'export output_dir_top=${cod_dir}/output/06.2-cod-RNAseq-alignment-genome'
echo 'export raw_fastqc_dir=${cod_dir}/output/05-cod-RNAseq-trimming/raw-fastqc'
echo 'export raw_reads_dir=${cod_dir}/data/05-cod-RNAseq-trimming/raw-reads'
echo 'export trimmed_fastqc_dir=${cod_dir}/output/05-cod-RNAseq-trimming/trimmed-fastqc'
echo 'export trimmed_reads_dir=${cod_dir}/output/05-cod-RNAseq-trimming/trimmed-reads'
echo 'export hisat2_output_dir=${output_dir_top}/hisat2'
echo ""


echo "# Input/Output files"
echo 'export genome_fasta_dir=${cod_dir}/data'
echo 'export genome_fasta_name="GCF_031168955.1_ASM3116895v1_genomic"'
echo 'export genome_fasta="${genome_fasta_dir}/${genome_fasta_name}"'
echo 'export genome_gff_name="GCF_031168955.1_ASM3116895v1"'
echo 'export genome_gff="${genome_gff_dir}/${genome_gff_name}"'
echo 'export hisat2_exon_name="G_macrocephalus_exon.tab"'
echo 'export hisat2_exon="${hisat2_output_dir}/${hisat_exon_name}"'
echo 'export hisat2_splice_sites_name="G_macrocephalus_splice_sites.tab"'
echo 'export hisat2_splice_sites="${hisat2_output_dir}/${hisat2_splice_sites_name}"'
echo 'export hisat2_index_name="G_macrocephalus_Hisat2_index.idx"'
echo 'export hisat2_index="${hisat2_output_dir}/${hisat2_index}"'


echo "# External data URLs and checksums"
echo 'export genome_fasta_url="https://owl.fish.washington.edu/halfshell/genomic-databank/GCF_031168955.1_ASM3116895v1_genomic.fna"'
echo 'export genome_fasta_checksum="5144890d4eceb0b258d92db3f35c681e"'
echo 'export genome_gff_url="https://owl.fish.washington.edu/halfshell/genomic-databank/GCF_031168955.1_ASM3116895v1.gff"'
echo 'export genome_gff_checksum="173fb3c159e474391c5c4aa1f7230024"'


echo "# Paths to programs"
echo 'export hisat2_exons=/home/shared/hisat2-2.2.1/hisat2_extract_exons.py'
echo 'export hisat2_splice_sites=/home/shared/hisat2-2.2.1/hisat2_extract_splice_sites.py'
echo 'export hisat2_build=/home/shared/hisat2-2.2.1/hisat2-build'
echo ""


echo "# Set number of CPUs to use"
echo 'export threads=20'
echo ""


echo "# Programs associative array"
echo "declare -A programs_array"
echo "programs_array=("
echo '[hisat2_exons]="${hisat2_exons}" \'
echo '[hisat2_splice_sites]="${hisat2_splice_sites}" \'
echo '[hisat2_build]="${hisat2_build}" \'
echo '[trinity_abund_to_matrix]="${trinity_abund_to_matrix}" \'
echo ")"
} > .bashvars

cat .bashvars
```

# Align to reference genome (Hisat2)

## Retrieving the reference genome and gff

```{r download-genome-fasta, engine='bash', eval=FALSE}
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${genome_fasta_dir} \
--recursive \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
--accept "${genome_fasta_name}.fna" ${genome_fasta_url}
```

```{r download-genome-gff, engine='bash', eval=FALSE}
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${genome_gff_dir} \
--recursive \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
--accept "${genome_gff_name}.gff" ${genome_gff_url}
```

```{r check-genome-dir, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

ls -lh "${genome_fasta_dir}"
```


## Verify genome FastA MD5 checksum

```{r verify-genome-fasta-checksum, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

cd "${genome_fasta_dir}"

# Checksums file contains other files, so this just looks for the sRNAseq files.
md5sum --check <<< "${genome_fasta_checksum}  ${genome_fasta_name}.fna"

```

```{r verify-genome-gff-checksum, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

cd "${genome_fasta_dir}"

# Checksums file contains other files, so this just looks for the sRNAseq files.
md5sum --check <<< "${genome_gff_checksum}  ${genome_gff_name}.fna"

```

## Building Index

```{r create-exons-tab-file, engine='bash'}
# Create Hisat2 exons tab file
"${programs_array[hisat2_exons]}" \
"${genome_gff}.gff" \
> "${exons}"
```

``` {r create-splice-sites-tab-file, engine='bash'}
# Create Hisat2 splice sites tab file
"${programs_array[hisat2_splice_sites]}" \
"${transcripts_gtf}" \
> "${splice_sites}"
```

```{r build-hisat2-index, engine='bash'}
# Build Hisat2 reference index using splice sites and exons
"${programs_array[hisat2_build]}" \
"${genome_fasta}.fna" \
"${genome_index_name}" \
--exon "${exons}" \
--ss "${splice_sites}" \
-p "${threads}" \
2> hisat2-build_stats.txt
```

```{r check-index-file, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

ls -lh ${hisat2_output_dir}
```





## Sample Quantification

Hisat2 can run quantification on either single- or paired-end reads. The default option is paired-end, which requires the input of an even number of paired fastq files (e.g., pairA_R1.fastq, pairA_R2.fastq). 
To use single-end mode, include the --single flag, as well as -l (--fragment-length=DOUBLE, estimated avg. fragment length) and -s (--sd=DOUBLE, estimates stand. dev. of fragment length), and a number of fastq files.
Again, gzipped files are acceptable.

Hisat2 quant is rather finicky about how you input sets of paired reads, and you can only input a single pair at a time. To circumvent, I'll create a quantification function and apply it iteratively to each pair of reads using a loop.

```{r hisat2-quantification, engine='bash', eval=FALSE}
# Load bash variables into memory
source .bashvars

# Function to run Hisat2 quant. Takes two (paired) reads as input, outputs to sample-associated directory
run_hisat2_quant() {
    source .bashvars  # Source .bashvars inside the function to make its variables accessible
    local R1_fastq=${1}
    local R2_fastq=${2}
    
    cd ${hisat2_output_dir}
    sample_num=$(basename "${R1_fastq}" "_R1_001.fastq.gz")
    mkdir hisat2_quant_${sample_num}

    ${programs_array[hisat2]} quant \
        --threads=${threads} \
        --index="${hisat2_output_dir}/${hisat2_index_name}" \
        --output-dir="${hisat2_output_dir}/hisat2_quant_${sample_num}" \
        --bootstrap-samples=100 \
        ${raw_reads_dir}/${R1_fastq} ${raw_reads_dir}/${R2_fastq}
}



# Iteratively apply run_hisat2_quant on each pair of input reads
for file_r1 in "${raw_reads_dir}"/*_R1_001.fastq.gz; do
    # Extract the sample name from the file name
    sample_name=$(basename "${file_r1}" "_R1_001.fastq.gz")

    # Form the file names (function takes input file names, not paths)
    file_r1_name="${sample_name}_R1_001.fastq.gz"
    file_r2_name="${sample_name}_R2_001.fastq.gz"

    # Check that the sample hasn't already been quantified
    if [ ! -d "${hisat2_output_dir}/hisat2_quant_${sample_name}" ]; then
    
        # Check if the corresponding R2 file exists
        if [ -e "${raw_reads_dir}/${file_r2}" ]; then
            # Run Hisat2 quant on the file pair
            run_hisat2_quant "${file_r1_name}" "${file_r2_name}" 

            echo "Processed sample: ${sample_name}"
        fi
    else
        echo "Sample already processed: ${sample_name}"
    fi
done
```

Check that we have the appropriate number of output folders. We should have (#rawreads/2)+1 -- one output folder for each pair of reads plus the single Hisat2 index file
```{r check-hisat2-output, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

# Count number of raw read files
cd ${raw_reads_dir}
echo "Number of raw reads:"
ls -1 | wc -l

# Count number of Hisat2 output 
cd ${hisat2_output_dir}
echo "Number of output files/folders"
ls -1 | wc -l
```


## Trinity Matrix with Hisat2 Output

```{r hisat2-trinity-matrix, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

cd ${hisat2_output_dir}

${programs_array[trinity_abund_to_matrix]} \
--est_method 'Hisat2' \
--gene_trans_map 'none' \
--out_prefix 'Hisat2' \
--name_sample_by_basedir ${hisat2_output_dir}/hisat2_quant_*/abundance.tsv

ls -lh ${hisat2_output_dir}
```
