---
title: "13.0.0-RNAseq-edgeR"
author: "Sam White"
date: "2024-05-29"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    toc: true
    number_sections: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
link-citations: true
---

```{r setup, include=FALSE}
library(knitr)
library(dplyr)
library(edgeR)
library(ggplot2)
library(ghibli)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

# EDGER GROUPING VECTOR

## Check sample info

Only looking at 9<sup>o</sup>C and 16<sup>o</sup>C
```{r check-sample-info, engine='bash', eval=TRUE}
echo "Header for ../data/DESeq2_Sample_Information.csv:"
echo ""

head -n 1 ../data/DESeq2_Sample_Information.csv

echo ""
echo "---------------------------------------------"
echo ""
awk -F"," '$4 == 16 || $4 == 9 {print $2, $4}' ../data/DESeq2_Sample_Information.csv | sort -n -k1,1
```

## Create edgeR grouping vector

Since all sample IDs `<=` to `36` are part of the 16<sup>o</sup>C treatment, we can use this to create vector which matches sample ordering in `../output/10.1-hisat-deseq2/gene_count_matrix.csv`

```{r edgeR-grouping, eval=TRUE}
# Read the first line of the CSV file
header <- readLines("../output/10.1-hisat-deseq2/gene_count_matrix.csv", n = 1)

# Split the header to extract the values
values <- as.numeric(unlist(strsplit(header, ","))[-1]) # Remove the first element which is "gene_id"

# Apply the conditional logic to generate the temperatures vector
temperatures <- ifelse(values <= 36, "16C", "9C")

# Print the resulting vector
print(temperatures)
```


# DEGs

## Load count matrix

### Reformat header

```{r reformat-counts-header, eval=TRUE}
# Read the entire CSV file
csv_file <- readLines("../output/10.1-hisat-deseq2/gene_count_matrix.csv")

# Extract the header line
header <- csv_file[1]

# Split the header to extract the values
header_values <- unlist(strsplit(header, ","))

# Prepend 'sample_' to each value (excluding 'gene_id')
header_values[-1] <- paste0("sample_", header_values[-1])

# Combine the modified header back into a single string
modified_header <- paste(header_values, collapse = ",")

# Replace the old header with the new modified header in the CSV content
csv_file[1] <- modified_header

# Convert the modified CSV content to a data frame
csv_content <- read.csv(textConnection(csv_file), row.names = 1)

```

### Convert to matrix
```{r convert-gene-counts-to-matrix, eval=TRUE}
# Convert the data frame to a matrix
gene_count_matrix <- as.matrix(csv_content)

# Print the matrix
print(head(gene_count_matrix))

```


### Load matrix into edgeR
```{r load-gene-count-matrix, eval=TRUE}
dge <- DGEList(counts = gene_count_matrix, group = factor(temperatures))

dge

```
### Filter low counts

Filters for genes with at >= `10` reads across at least 3 samples.
```{r filter-low-gene-counts}
keep <- filterByExpr(dge)

dge <- dge[keep, , keep.lib.sizes=FALSE]

dge
```
### Normalization
```{r normalize-gene-read-counts}
dge <- calcNormFactors(object = dge)

dge
```

### Model fitting and estimating dispersions
```{r model-dispersion-genes}
dge <- estimateDisp(dge)

dge
```

### Exact test (Gene expression)
```{r }
exact_test_genes <- exactTest(dge)

exact_test_genes

```

### Differential gene expression summary
```{r }
summary(decideTests(object = exact_test_genes, p.value = 0.05))
```

### Peek at DEGs
```{r }
top_degs_table <-  topTags(object = exact_test_genes, n = "Inf", adjust.method = "fdr", p.value = 0.05)$table
top_degs_table

```

### Volcano plot
```{r }
ghibli_colors <- ghibli_palette("PonyoMedium", type = "discrete")

ghibli_subset <- c(ghibli_colors[3], ghibli_colors[6], ghibli_colors[4])

# Create new column and fill with `NA`
top_degs_table$topDE <- "NA"

# Set value of "Up" for genes with FC > 1 and FDR < 0.05
top_degs_table$topDE[top_degs_table$logFC > 1 & top_degs_table$FDR < 0.05] <- "Up"

# Set value of "Down" for genes with FC < -1 and FDR < 0.05
top_degs_table$topDE[top_degs_table$logFC < -1 & top_degs_table$FDR < 0.05] <- "Down"

ggplot(data=top_degs_table, aes(x=logFC, y=-log10(FDR), color = topDE)) + 
  geom_point() +
  theme_minimal() +
  scale_colour_discrete(type = ghibli_subset, breaks = c("Up", "Down"))
```